### PR TITLE
Content server auto scale

### LINKED ISSUE #19: add auto scaling for content server tasks
Content servers act as HTTP server for serving the file content. Each task container has maximum capability to process the data transferring. We can use auto scaling capability to increase the total throughput of content servers and cost optimization.

### Use Case

- [ ] add auto scaling for content server task
- [ ] unit tests

Based on the content server nature, the auto scaling should based on network traffic. Below are some criteria we should consider when designing the metrics for auto scaling, 

- fargate instance network capability
- efs client network throughput, currently maximum throughout is 500MB/s per client(https://docs.aws.amazon.com/efs/latest/ug/limits.html#limits-client-specific)

### Other

* [ ] :wave: I may be able to implement this feature request

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by jiegec**
To close #19 

- [x] run cloudwatch agent in content server
- [x] collect metrics
- [x] add auto scaling for content server task
- [x] unit tests
- [x] update readme

### CODE REVIEW

File: lib/content-server.ts [hunk 1]
```diff
@@ -113,25 +113,49 @@ export class ContentServerStack extends cdk.NestedStack {
...
+        });
+        scaling.scaleOnMetric('NetworkScaling', {
+            metric: new cloudwatch.Metric({
+                namespace: 'OpenTuna',
+                metricName: 'net_bytes_sent',
+                dimensions: {
+                    interface: "eth1"
+                }
+            }),
+            scalingSteps: [{
```
Comment (zxkane) [inline lib/content-server.ts:145]: Let's test the network performance of Fargate in ZHY region, then decide how to use metrics to scaling out/in.

File: lib/content-server.ts [hunk 2]
```diff
@@ -113,25 +113,49 @@ export class ContentServerStack extends cdk.NestedStack {
...
+        const scaling = service.autoScaleTaskCount({
+            minCapacity: 1,
+            maxCapacity: 16
+        });
+        scaling.scaleOnMetric('NetworkScaling', {
+            metric: new cloudwatch.Metric({
+                namespace: 'OpenTuna',
+                metricName: 'net_bytes_sent',
+                dimensions: {
+                    interface: "eth1"
```
Comment (zxkane) [inline lib/content-server.ts:142]: Might aggregate all network interfaces.

File: lib/content-server.ts [hunk 3]
```diff
@@ -43,68 +81,30 @@ export class ContentServerStack extends cdk.NestedStack {
...
+        });
+        const cloudWatchAgent = taskDefinition.addContainer("cloudwatch-agent", {
+            image: ecs.ContainerImage.fromRegistry('amazon/cloudwatch-agent:latest'),
+            logging: new ecs.AwsLogDriver({
+                streamPrefix: usage,
+                logGroup: cloudWatchAgentlogGroup,
             }),
+            essential: false,
+            secrets: {
+                "CW_CONFIG_CONTENT": ecs.Secret.fromSsmParameter(param),
```
[the remark this answers is not in the document]
Comment (jiegec) [reply lib/content-server.ts:106]: It is a configmap in fact, although it is put under `ecs.Secret`

File: lib/content-server.ts (hunk 2 above)
Comment (jiegec) [reply lib/content-server.ts:143]: Seems that `For metrics produced by certain AWS services, such as Amazon EC2, CloudWatch can aggregate data across dimensions. For example, if you search for metrics in the AWS/EC2 namespace but do not specify any dimensions, CloudWatch aggregates all data for the specified metric to create the statistic that you requested. CloudWatch does not aggregate across dimensions for your custom metrics.`

File: lib/content-server.ts (hunk 1 above)
Comment (zxkane) [reply lib/content-server.ts:168]: why not using percentage for scaling?

**Review by zxkane — CHANGES_REQUESTED** [review]

File: lib/content-server.ts [hunk 4]
```diff
@@ -43,95 +82,79 @@ export class ContentServerStack extends cdk.NestedStack {
...
+        });
+        // each task instance has about 500Mbps bandwidth
+        // 500Mbps * 60s/min / 8b/B = 3750MB/min
+        // must use direct metric because of api limitation
+        // so only eth1 is used
+        scaling.scaleToTrackCustomMetric('NetworkBandwidthScaling', {
+            metric: bytesSentEth1,
+            targetValue: 3 * 1024 * 1024 * 1024, // 3GiB/min
+            scaleInCooldown: cdk.Duration.minutes(10),
+            scaleOutCooldown: cdk.Duration.minutes(10),
```
Comment (zxkane) [inline lib/content-server.ts:155]: let's use 3min for scaling out

**Review by zxkane — CHANGES_REQUESTED** [review]

File: lib/content-server.ts [hunk 5]
```diff
@@ -43,95 +82,79 @@ export class ContentServerStack extends cdk.NestedStack {
...
+            dimensions: {
+                interface: "eth1"
+            },
+            statistic: cloudwatch.Statistic.AVERAGE,
+        });
+        // each task instance has about 500Mbps bandwidth
+        // 500Mbps * 60s/min / 8b/B = 3750MB/min
+        // must use direct metric because of api limitation
+        // so only eth1 is used
+        scaling.scaleToTrackCustomMetric('NetworkBandwidthScaling', {
```
Comment (zxkane) [inline lib/content-server.ts:151]: Per doc https://docs.aws.amazon.com/autoscaling/application/userguide/application-auto-scaling-target-tracking.html, the auto scaling support more than one policy.

I prefer to add another policy to more smoothly scale in,

1. set the metric of 50% of maximum traffic
1. set 10m as cool down period of scale in
1. set a super large period of cool down of scale out

Those two policies will make the scaling out when the metrics exceeding 3GB metrics threshold. And the scaling in must the traffic is lower than 50% utilities.

**Review by zxkane — APPROVED** [review]

**Pull request merged**